index.d.ts

```diff
@@ -1,288 +1,473 @@
-    columns: HeaderColumn<D>[]
-    state: State<D>
-    debug?: boolean
-    loading: boolean
-    defaultColumn?: Partial<Column<D>>
-  }
-
-  // The empty definition of TableState is not an error. It provides a definition
-  // for the state, that can then be extended in the users code.
-  //
-  // e.g.
-  //
-  // export interface TableState<D = {}}>
-  //     extends UsePaginationState,
-  //       UseGroupByState,
-  //       UseSortbyState<D>,
-  //       UseFiltersState<D> {}
-
-  // eslint-disable-next-line @typescript-eslint/no-empty-interface
-  export interface TableState<D = {}> {}
-
-  export type SetState<D> = (
-    updater: (old: TableState<D>) => TableState<D>,
-    actions: any
-  ) => void
+  export interface UseRowSelectState<D extends object> {
+    selectedRows: IdType<D>[]
+    manualRowSelectedKey: IdType<D>
+  }
 
-  export type State<D> = [TableState<D>, SetState<D>]
+  export interface UseRowSelectInstanceProps<D extends object> {
+    toggleRowSelected: (rowPath: IdType<D>, set?: boolean) => void
+    toggleRowSelectedAll: (set?: boolean) => void
+    getToggleAllRowsSelectedProps: (props: object) => object
+    isAllRowsSelected: boolean
+  }
 
-  export function useTable<D = {}>(
-    props: TableOptions<D>,
-    ...plugins: any[]
-  ): TableInstance<D>
+  export interface UseRowSelectRowProps<D extends object> {
+    isSelected: boolean
+    toggleRowSelected: (set?: boolean) => void
+  }
+  /* #endregion */
+  
+  /* #region useRowState */
+  export interface useRowState<D extends object = {}> extends PluginHook<D> {}
+
+  export type UseRowStateOptions<D extends object> = Partial<{
+    initialRowStateAccessor: (row: Row<D>) => object
+  }>
+  
+  export interface UseRowStateState<D extends object> {
+    rowState: Partial<{ cellState: UseRowStateLocalState<D>, rowState: UseRowStateLocalState<D> }>
+  }
 
-  export function useFilters<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & {
-    rows: Row<D>[]
+  export interface UseRowStateInstanceProps<D extends object> {
+    setRowState: (rowPath: string[], updater: UseRowUpdater) => void
+    setCellState: (rowPath: string[], columnID: IdType<D>, updater: UseRowUpdater) => void
   }
 
-  export function useSortBy<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & {
-    rows: Row<D>[]
+  export interface UseRowStateRowProps<D extends object> {
+    state: unknown
+    setState: (updater: UseRowUpdater) => void
+  }
+  export interface UseRowStateCellProps<D extends object> {
+    state: UseRowStateLocalState<D>
+    setState: (updater: UseRowUpdater) => void
   }
 
-  export function useGroupBy<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & { rows: Row<D>[] }
+  type UseRowUpdater<T = unknown> = T | ((prev: T) => T)
+  type UseRowStateLocalState<D extends object, T = unknown> = Record<IdType<D>, T>
+  /* #endregion */
+  
+  /* #region useSortBy */
+  export interface useSortBy<D extends object = {}> extends PluginHook<D> {}
```

Comment:
This should probably be `export function useSortBy<D extends object = {}>(): PluginHook<D>;` for value usage? Same or other hooks.

Reply:
Good catch, fixed.